109 adapt status line if cgi script provides one

u413-284-si opened the pull request:
## Overview
1. Add parsing of received response body
2. Add utility functions to extract status code from a string
3. Include parsed headers into built response
4. Fix integration tests


### re 1.
- add one main function parseResponseBody() to be called in buildResponse() before the response is constructed
- subfunction parseResponseStatusLine() to extract the status code if given: two indicators for status line allowed
1. Status: such as _Status 200 OK_
2. HTTP/1.1: such as _HTTP/1.1 200 OK_
- subfunction parseResponseHeaders() to extract all headers provided in the received response (CGI)
- headers are saved in a map for further usage
- subfunction processResponseHeaders() to check if current Connection is required to be closed by the CGI script
- further processing could be added if necessary

### re 2. 
- new functions extractStatusCode() and stringToStatusCode()

### re 3.
- only the body of a CGI response will now be added to the final response
- removed check in appendStatusLine()
- appendHeadersCGI() merged with appendHeaders()
- cross-check whether certain headers have been provided by the CGI response: if yes, use these, otherwise construct the headers
- all CGI headers are added to the response header
- resetBuilder() also clears the map of response headers

### re 4.
- activate all CGI integration tests
- fix CGI POST tests by allowing POST requests in the config location

closes #109

gwolf-011235 commented on test/integration/server_response/test_CGI.py:
```diff
@@ -7,41 +7,41 @@ def test_no_CGI_defined():
+          "but they would soon find out."
+    )
+    response = requests.post("http://127.0.0.1:8080/cgi-bin/upperCase.sh", data=payload)
+    assert response.status_code == 200
```
This should check if the response is actually uppercase:
`assert payload.upper() in response.text`

u413-284-si replied:
Added

u413-284-si commented on inc/ResponseBuilder.hpp:
```diff
@@ -31,10 +34,15 @@ class ResponseBuilder {
 	void appendHeadersCGI(const HTTPRequest& request);
 	std::string getMIMEType(const std::string& extension);
 	void initMIMETypes();
+	void parseResponseBody(HTTPRequest& request);
+	void parseResponseStatusLine(HTTPRequest& request);
+	void parseResponseHeaders();
+	void processResponseHeaders(HTTPRequest& request);
```
Renamed and moved

u413-284-si commented on inc/utilities.hpp:
```diff
@@ -28,6 +28,8 @@ std::string getGMTString(time_t now, const std::string& format);
 
 std::string statusCodeToReasonPhrase(statusCode status);
+statusCode stringToStatusCode(std::string& str);
+statusCode extractStatusCode(const std::string& statusLine);
```
Added

u413-284-si commented on inc/ResponseBuilder.hpp:
```diff
@@ -31,10 +34,15 @@ class ResponseBuilder {
 	const FileSystemPolicy& m_fileSystemPolicy;
 	std::stringstream m_responseHeader;
 	std::string m_responseBody;
+	std::map<std::string, std::string> m_responseHeaders;
```
Renamed

u413-284-si commented on cgi-bin/time.py:
```diff
@@ -2,8 +2,11 @@
+print("Connection: close\r\n", end="")
+print("Cache-Control: no-cache\r\n", end="")
+print("Pragma: no-cache\r\n\r\n", end="")
 print("<html>")
 print("<head>")
 print(datetime.datetime.strftime(datetime.datetime.now(), "<h1>  %H:%M:%S </h1>"))
```
Done

u413-284-si commented on src/ResponseBuilder.cpp:
```diff
@@ -101,54 +98,48 @@ void ResponseBuilder::appendStatusLine(const HTTPRequest& request)
-	// Content-Length
-		m_responseHeader << "Content-Length: " << m_responseBody.length() << "\r\n";
+		// Content-Type
+		if (m_responseHeaders.find("Content-Type") != m_responseHeaders.end()) {
```
Correct, added reusable function to check for header with iterator

u413-284-si commented on src/ResponseBuilder.cpp:
```diff
@@ -101,54 +98,48 @@ void ResponseBuilder::appendStatusLine(const HTTPRequest& request)
+			m_responseHeader << "Content-Length: " << m_responseBody.length() << "\r\n";
 	}
+
+	// Various headers from response
```
Checking each header. Would stick with appendHeaders() as sole function as we would have to anyways add those headers which are not provided by CGI. Easier logic

gwolf-011235 replied:
But then we treat the special case as normal: the only header which is added outside of ResponseBuilder::appendHeaders() for all non-CGI requests is "Location". This means we _know_ all checks for other headers will result in false - but we do them anyways.
A simple bool check and extra function for special case seems better to me.

u413-284-si replied:
This results in two function where appendHeaders():
```
void ResponseBuilder::appendHeaders(const HTTPRequest& request)
{
	if (!m_responseBody.empty()) {
		// Content-Type
		m_responseHeaderStream << "Content-Type: "
								   << getMIMEType(webutils::getFileExtension(request.targetResource)) << "\r\n";
		// Content-Length
		m_responseHeaderStream << "Content-Length: " << m_responseBody.length() << "\r\n";
	}

	// Server
	m_responseHeaderStream << "Server: TriHard\r\n";

	// Date
	m_responseHeaderStream << "Date: " << webutils::getGMTString(time(0), "%a, %d %b %Y %H:%M:%S GMT") << "\r\n";

	// Location
	std::map<std::string, std::string>::const_iterator iter = request.headers.find("location");
		if (iter != request.headers.end()) {
			m_responseHeaderStream << "Location: " << iter->second << "\r\n";
		}
	
	// Connection
	if (request.shallCloseConnection)
		m_responseHeaderStream << "Connection: close\r\n";
	else
		m_responseHeaderStream << "Connection: keep-alive\r\n";

	// Delimiter
	m_responseHeaderStream << "\r\n";
}
```

And then we would still need the exact same function as before as appendCGIHeaders() as we don't know which header is actually provided by the CGI response, so we need to be able to create the header ourselves if not provided:
```
void ResponseBuilder::appendHeaders(const HTTPRequest& request)
{
	if (!m_responseBody.empty()) {
		// Content-Type
		if (!hasCGIHeader("Content-Type"))
			m_responseHeaderStream << "Content-Type: "
								   << getMIMEType(webutils::getFileExtension(request.targetResource)) << "\r\n";
		// Content-Length
		if (!hasCGIHeader("Content-Length"))
			m_responseHeaderStream << "Content-Length: " << m_responseBody.length() << "\r\n";
	}

	// Various headers from response
	for (std::map<std::string, std::string>::const_iterator iter = m_responseHeaders.begin();
		 iter != m_responseHeaders.end(); ++iter)
		m_responseHeaderStream << iter->first << ": " << iter->second << "\r\n";

	// Server
	if (!hasCGIHeader("Server"))
		m_responseHeaderStream << "Server: TriHard\r\n";

	// Date
	if (!hasCGIHeader("Date"))
		m_responseHeaderStream << "Date: " << webutils::getGMTString(time(0), "%a, %d %b %Y %H:%M:%S GMT") << "\r\n";

	// Location
	if (!hasCGIHeader("Location")) {
		std::map<std::string, std::string>::const_iterator iter = request.headers.find("location");
		if (iter != request.headers.end()) {
			m_responseHeaderStream << "Location: " << iter->second << "\r\n";
		}
	}

	// Connection
	if (request.shallCloseConnection)
		m_responseHeaderStream << "Connection: close\r\n";
	else
		m_responseHeaderStream << "Connection: keep-alive\r\n";

	// Delimiter
	m_responseHeaderStream << "\r\n";
}
```

The latter one would cover both cases. So the differentiation only serves for a reduced cost in execution by avoiding the checks. On the other hand it creates not exactly code "duplication", but it is very similar. Debatable which is worse, changed anyways.

gwolf-011235 replied:
I agree it doesn't look great. In my mind this is a path every request takes, and it looks like an easy performance gain.

But since we have m_responseHeaders anyway: we could init "standard" headers in the map in the constructor similar to initMIMETypes():
```
void ResponseBuilder::initStandardHeaders()
{
  m_responseHeaders["status"] = "OK";
  m_responseHeaders["content-length"] = "0";
  ... // other headers we always need
}
```

Then we don't need to search them, since we know they exist. ResponseBodyBuilder would just overwrite the headers with correct values. Likewise CGI header parsing would just overwrite existing headers, if provided by script. 
After we are done we reset their values, so we don't even have to add/remove them everytime.

Additional headers don't fit nicely in this scheme though. If we loop the map, check if header is a standard header and if not we print, then we have unnecessary check all over again. I think more performant would be another map variable: if it's empty we don't do anything. If CGI provided extra headers we can just loop through. We still would need the check for standard headers, but we only need it while CGI header parsing.

u413-284-si replied:
IMO this creates more complexity w/o much of a gain. We would still have two functions and an additional map. Would leave it as is.

gwolf-011235 replied:
Nah, we have only one function which just constructs the header part out of m_responseHeaders (and m_responseSpecialHeaders). Would also eliminate appendStatusLine():

```
void ResponseBuilder::appendHeaders(const HTTPRequest& request)
{
  m_responseHeaderStream << "HTTP/1.1 " << m_responseHeaders["status"] << "\r\n";

  if (!m_responseBody.empty()) {
    // Content-Type
    m_responseHeaderStream << "Content-Type: " << m_responseHeaders["content-type"] << "\r\n";
    // Content-Length
    m_responseHeaderStream << "Content-Length: " << m_responseHeaders["content-length"] << "\r\n";
    }

  // Server
  m_responseHeaderStream << "Server: " << m_responseHeaders["server"] << "\r\n";

  // Date
  m_responseHeaderStream << "Date: " << m_responseHeaders["date"] << "\r\n";

  // Location
  std::map<std::string, std::string>::const_iterator iter = m_responseHeaders.find("location");
    if (iter != request.headers.end()) {
      m_responseHeaderStream << "Location: " << iter->second << "\r\n";
    }
	
  // Connection
  m_responseHeaderStream << "Connection: " << m_responseHeaders["connection"] << "\r\n";

  // Various headers from CGI response, does nothing for all other things.
  for (std::map<std::string, std::string>::const_iterator iter = m_responseSpecialHeaders.begin();
   iter != m_responseSpecialHeaders.end(); ++iter)
    m_responseHeaderStream << iter->first << ": " << iter->second << "\r\n";

  // Delimiter
  m_responseHeaderStream << "\r\n";
}
```

u413-284-si replied:
Ok, but we still need the extra function to initialize the map. And the parsing of the CGI response header would become more complex as you would have to check whether the header belongs to the "standard" map or to the special one and save accordingly, so for each header the standard map would need to be traversed. I still don't think it is worth the rewrite.

gwolf-011235 replied:
> Ok, but we still need the extra function to initialize the map.

True.

> so for each header the standard map would need to be traversed.

I would just loop through the six standard header name strings after extracting the header name. If hit > use standard, if not use special map. Should be very fast. 

We can also leave it.

u413-284-si commented on src/ResponseBuilder.cpp:
```diff
@@ -185,3 +176,114 @@ std::string ResponseBuilder::getMIMEType(const std::string& extension)
+
+void ResponseBuilder::parseResponseBody(HTTPRequest& request)
+{
+	if (m_responseBody.empty())
```
Removed

u413-284-si commented on src/ResponseBuilder.cpp:
```diff
@@ -185,3 +176,114 @@ std::string ResponseBuilder::getMIMEType(const std::string& extension)
 	return m_mimeTypes.at("default");
 }
+
+void ResponseBuilder::parseResponseBody(HTTPRequest& request)
```
Not necessary anymore, removed parseResponseStatusLine()

u413-284-si commented on src/ResponseBuilder.cpp:
```diff
@@ -185,3 +176,114 @@ std::string ResponseBuilder::getMIMEType(const std::string& extension)
+ *
+ * @param request The HTTPRequest object to be updated with the parsed status code.
+ */
+void ResponseBuilder::parseResponseStatusLine(HTTPRequest& request)
```
Indeed, I was under the impression that also normal status lines à la  "HTTP/1.1 200 OK" would be acceptable from the CGI script. But the RFC is quite clear about it. Removed parseResponseStatusLine() and treat "Status" as another header being parsed in parseCGIResponseHeaders().

u413-284-si commented on src/ResponseBuilder.cpp:
```diff
@@ -185,3 +176,114 @@ std::string ResponseBuilder::getMIMEType(const std::string& extension)
+ *
+ * @param request The HTTPRequest object to be updated with the processed headers.
+ */
+void ResponseBuilder::processResponseHeaders(HTTPRequest& request)
```
The logic of handling the connection header is a bit more complex than the other headers - setting the boolean in addition - so it shall still remain. Also the handling of the status header happens here now as well.

gwolf-011235 replied:
I think the complex logic is only for the boolean? Since we have special function appendCGIheader() we can just use the same logic as with other headers:
```
// Connection
if (!hasCGIHeader("connection"))
		m_responseHeaderStream << "Connection: keep-alive\r\n"; 
```

Every HTTP/1.1 is keep alive by default. Since we have no error (we wouldn't enter the function in that case), we can just search for the header and if not use keep-alive.

Same can be done for "Status", but then if (request.hasCGI) would also need to guard against using appendStatusLine(). 

Status header interestingly has a validation logic inside of extractStatusCode() while Connection header does not (we accept any value but only act on "close"). Since Connection header accepts only two values (close / keep-alive) I think a validateResponseHeaders() would make more sense?

u413-284-si replied:
But we do enter the function in case of an error. If before anything would go wrong while processing the CGI request, the boolean shallCloseConnection would be set to true, we would create our error body and then enter this function to create the headers. We would have no CGIHeaders as we didn't enter the CGI parsing part, but we would still need to ouput <Connection: close>. So the logic of the other headers cannot be applied here.

The Status header requires handling of the potential error which was added. Also added validation of Connection header to only accept keep-alive or close. The function is renamed to validateResponseHeaders().

gwolf-011235 replied:
Why would we enter? 

If we have an error while processing the CGI request we shouldn't even try to parse anything, since we don't know what state we have. Or do you want to parse the headers in any case? And then send a error response with maybe parsed headers of CGI?
I thought we guard like we had before `if (request.hasCGI && request.httpStatus == StatusOK)`

gwolf-011235 replied:
I guess we want to enter if script contains a Status: and that status is an error code. 

We could not set the request.httpstatus. The appendCGIHeader() function would then construct the status line with the Status (if not given use StatusOK). Then only succesful CGI request (everything went ok) would enter appendCGIHeader()

u413-284-si replied:
Then we would have to guard against using appendStatusLine(). 
Don't you think that this logic is cumbersome? It requires quite some mental power to follow all the cases when what output would be printed diminishing the readability.

I think it is more straightforward to just take the value provided (status code or connection:keep/close) and apply it to the variables which are anyways there and used by the rest of the system. In the case of connection it is even obligatory, otherwise we would be ignoring the "command". So why not use the available variables and print out what state they are currently in? No need to keep track of multiple potentially differing statements (one in the header, another in the variable).

gwolf-011235 replied:
Yes I think the logic is cumbersome. 
Which is why I suggested to use the newly introduced m_responseHeaders as standard for all requests.

Right now we pass a generic named variable m_responseHeaders, which is only used for (successful) CGI requests. We also validate and interpret some of the cgi headers but not all of them, because we have existing variables. It all feels very patchworky, like an afterthought.

We could at least find middle ground to reintroduce special function appendCGIHeaders() so that CGI stuff = "search for possible CGIHeaders" is isolated from common control flow. 
For the same reason I am against routing error flow of CGI requests to appendCGIHeaders() since they don't differ from other (error) responses > they won't have CGI headers. 

> In the case of connection it is even obligatory, otherwise we would be ignoring the "command". 

I don't understand?

> No need to keep track of multiple potentially differing statements (one in the header, another in the variable).

We have a specialised function appendCGIHeaders() which can do specialised things. We only go in when the CGI request is succesful.

Maybe I'm also thinking too complicated. What about a trio conference?

gwolf-011235 replied:
Disregard. I think you're right.

I was too concentrated on response building, that I forgot the main function of the bool: actually closing the connection. So yes we need to process it everytime.

I also reconsidered my stance on the dual function solution. My proposal is micro optimisation at best, which might not even faster: Branch prediction could help us out. So I would be down to eliminate appendCGIHeaders().

I instead propose that we add the location header to m_responseHeader in ResponseBodyBuilder which currently is added to the **request** headers at
- ResponseBodyHandler.cpp:96 (processing POST request)
- Server.cpp:994-995 (check if isRedirectionStatus) 

For that we could rename `hasCGIHeader` to `checkForExistingHeader` (as we discussed has is only for simple check which does nothing). Then doxygen could explain that currently only location header is added somewhere else.

One thing I had in my mind when I read about the Status header is that we could sent custom reason phrases with Status header. However we parse Status header and then use our implemented reason phrases, ignoring whats actually put in. 
Is it worth it to change?

u413-284-si replied:
Ok, let me repeat to see if I understand you correctly:
* we keep validateCGIResponseHeaders() to validate the _status_ and _connection_ header and to set the according variables
* we remove the differentiation between appendHeaders() and appendCGIHeaders() -> just one function which has all the checks (is that what you mean by eliminating appendCGIHeaders?):
```
if (!m_responseBody.empty()) {
		// Content-Type
		if (!hasCGIHeader("content-type"))
			m_responseHeaderStream << "Content-Type: "
								   << getMIMEType(webutils::getFileExtension(request.targetResource)) << "\r\n";
		// Content-Length
		if (!hasCGIHeader("content-length"))
			m_responseHeaderStream << "Content-Length: " << m_responseBody.length() << "\r\n";
	}

	// Server
	if (!hasCGIHeader("server"))
		m_responseHeaderStream << "Server: TriHard\r\n";

	// Date
	if (!hasCGIHeader("date"))
		m_responseHeaderStream << "Date: " << webutils::getGMTString(time(0), "%a, %d %b %Y %H:%M:%S GMT") << "\r\n";

	// Location
	if (!hasCGIHeader("location")) {
		std::map<std::string, std::string>::const_iterator iter = request.headers.find("location");
		if (iter != request.headers.end()) {
			m_responseHeaderStream << "Location: " << iter->second << "\r\n";
		}
	}

	// Various headers from response
	if (request.httpStatus < StatusMovedPermanently) {
		for (std::map<std::string, std::string>::const_iterator iter = m_responseHeaders.begin();
			iter != m_responseHeaders.end(); ++iter)
			m_responseHeaderStream << webutils::capitalizeWords(iter->first) << ": " << iter->second << "\r\n";
	}

	// Connection
	if (request.shallCloseConnection)
		m_responseHeaderStream << "Connection: close\r\n";
	else
		m_responseHeaderStream << "Connection: keep-alive\r\n";

	// Delimiter
	m_responseHeaderStream << "\r\n";
```
* rename hasCGIHeader() to checkForExistingHeader()
* add location header to m_responseHeader

Regarding the reason phrase, we could take the parsed value of _Status:_ and print it. Would be another check in appendStatusLine().

gwolf-011235 replied:
Yes to all (small clarification: Add location header only if it needs to be added).

Or we remove addStatusline entirely (it's just a one liner) and merge it with appendHeaders(). Then check for header Status.

Or we just not have custom reason phrase :D

u413-284-si replied:
Implemented discussed points, integrated status line in appendHeaders() and introduced a specific check in checkForExistingHeader() to allow for custom reason phrases

u413-284-si commented on src/ResponseBuilder.cpp:
```diff
@@ -185,3 +176,114 @@ std::string ResponseBuilder::getMIMEType(const std::string& extension)
+ * them into a map of header names and values. It updates the response headers map
+ * with the parsed headers.
+ */
+void ResponseBuilder::parseResponseHeaders()
```
I left those checks out because we are in control of the CGI scripts. We provide them and decide the headers with their respective formatting.

gwolf-011235 replied:
If we "ship" our server we are not in control anymore. I think at least webutils::lowercase would make usage easier (our consumer can write his headers capitalised if he wants).

u413-284-si replied:
Added lowercasing. Required the addition of a function to capitalize the words in the header, when we are printing them out again in our response.

gwolf-011235 replied:
Another point for my header solution :D

u413-284-si commented on test/integration/server_response/test_CGI.py:
```diff
@@ -7,41 +7,41 @@ def test_no_CGI_defined():
+def test_CGI_helloWorld():
+   print("Request for /cgi-bin/helloWorld.sh")
+   response = requests.get("http://127.0.0.1:8080/cgi-bin/helloWorld.sh")
+   assert response.status_code == 200
```
Changed

u413-284-si commented on test/integration/server_response/test_CGI.py:
```diff
@@ -7,41 +7,41 @@ def test_no_CGI_defined():
+def test_CGI_time():
+   print("Request for /cgi-bin/time.py")
+   response = requests.get("http://127.0.0.1:8081/cgi-bin/time.py")
+   assert response.status_code == 200
```
Added

u413-284-si commented on test/integration/server_response/test_CGI.py:
```diff
@@ -7,41 +7,41 @@ def test_no_CGI_defined():
-#     response = requests.post("http://127.0.0.1:8080/cgi-bin/upperCase.sh", data=payload)
-#     assert response.status_code == 200
+   response = requests.post("http://127.0.0.1:8081/cgi-bin/upload.py", params=query_params, data=payload)
+   assert response.status_code == 200
```
Changed

u413-284-si commented:
Added removal of fds whenever they are not required:
- clientFd removed when CGI is working, added after completion
- pipeToCGI only added if there is a body to be sent and removed after complete sending
- pipeFromCGI only added when status is _Connection::ReceiveFromCGI_, removed after complete receival

gwolf-011235 commented:
> Added removal of fds

Does it still work after removing the `if (activeFd != connection.m_clientFd) return;` ?

gwolf-011235 commented on inc/ResponseBuilder.hpp:
```diff
@@ -20,21 +23,22 @@
 	std::string getMIMEType(const std::string& extension);
 	void initMIMETypes();
 	void resetBuilder();
+	bool checkForCGIHeader(const std::string& headerName);
```
Rename to checkForExistingHeader()

u413-284-si replied:
Renamed

gwolf-011235 commented on inc/Server.hpp:
```diff
@@ -128,6 +129,7 @@ class Server {
 	std::vector<char> m_clientHeaderBuffer; /**< Buffer for reading request header */
 	std::vector<char> m_clientBodyBuffer; /**< Buffer for reading request body */
 	std::vector<char> m_cgiBodyBuffer; /**< Buffer for reading CGI response body */
+	std::map<std::string, std::string> m_responseHeaders; /**< Response headers */
```
Like above: I think we can move header functionality completely into ResponseBuilder / ResponseBodyHandler

u413-284-si replied:
Moved

gwolf-011235 commented on src/ResponseBodyHandler.cpp:
```diff
@@ -53,6 +55,7 @@ void ResponseBodyHandler::execute()
 			m_request.httpStatus = StatusInternalServerError;
 			handleErrorBody();
 		}
+		parseCGIResponseBody();
```
I think we need to adapt the check in the lines before this (53-57):
RFC 3875 Sect 6.3 [[LINK](https://datatracker.ietf.org/doc/html/rfc3875#section-6.3)] states: "At least one CGI field MUST be supplied". 
CGI field = Content-type | Location | Status

Since we now support all three (I think?) we need to revisit the check above. If one is present then is ok. 
However we can't just .find() for one of them since we need to restrict the search inside the header part (i.e. before \r\n\r\n) - currently is unrestricted search in m_request.body.

u413-284-si replied:
Changed, moved into parseCGIResponseHeaders() to use the header substring created therein

gwolf-011235 commented on src/ResponseBodyHandler.cpp:
```diff
@@ -108,6 +112,89 @@ void ResponseBodyHandler::execute()
+	std::map<std::string, std::string>::iterator iter = m_responseHeaders.find("status");
+	if (iter != m_responseHeaders.end()) {
+		m_request.httpStatus = extractStatusCode(iter->second);
+		if (m_request.httpStatus == StatusBadRequest) {
```
What if the CGI script set Status to StatusBadRequest? Could set to NoStatus

u413-284-si replied:
Can be set now

gwolf-011235 commented on src/ResponseBuilder.cpp:
```diff
@@ -40,20 +42,15 @@ void ResponseBuilder::buildResponse(Connection& connection)
-	ResponseBodyHandler responseBodyHandler(connection, m_responseBody, m_fileSystemPolicy);
+	ResponseBodyHandler responseBodyHandler(connection, m_responseBody, m_responseHeaders, m_fileSystemPolicy);
 	responseBodyHandler.execute();
+	appendHeaders(request);
```
I called it appendStatusLine and appendHeaders because we had one variable for header and body part.
I think a more apt name would now be constructResponseHeader or something?

u413-284-si replied:
Renamed to appendResponseHeader

gwolf-011235 commented on src/ResponseBuilder.cpp:
```diff
@@ -100,53 +84,71 @@ void ResponseBuilder::appendStatusLine(const HTTPRequest& request)
  */
 void ResponseBuilder::appendHeaders(const HTTPRequest& request)
 {
+	if (!checkForCGIHeader("status"))
```
Doxygen of appendHeaders needs update

u413-284-si replied:
Updated

gwolf-011235 reviewed (changes requested):
Looks good! Sorry for oscillating between suggestions.

I think we still have one RFC-non-conforming part, but I think should be easily fixable?

gwolf-011235 commented on src/ResponseBodyHandler.cpp:
```diff
@@ -108,6 +112,89 @@ void ResponseBodyHandler::execute()
+{
+	const size_t sizeCRLF = 2;
+	const size_t sizeCRLFCRLF = 4;
+	size_t posHeadersEnd = m_responseBody.find("\r\n\r\n");
```
Can be const

u413-284-si replied:
Done

gwolf-011235 commented on src/ResponseBodyHandler.cpp:
```diff
@@ -108,6 +112,89 @@ void ResponseBodyHandler::execute()
+ * This function processes the response headers and updates the HTTPRequest object
+ * with relevant information, such as whether the connection should be closed.
+ */
+void ResponseBodyHandler::validateCGIResponseHeaders()
```
RFC 3875 Sect 6.3 [LINK] states: "At least one CGI field MUST be supplied". And CGI field = Content-type | Location | Status

We have one check which only searches content-type, which can be removed if we do a check here (see above).

One of the three is sufficient, we can track with bool:
```
void ResponseBodyHandler::validateCGIResponseHeaders()
{
  bool hasCGIField = false;

  // Status
  std::map<std::string, std::string>::iterator iter = m_responseHeaders.find("status");
  if (iter != m_responseHeaders.end()) {
    hasCGIFIeld = true;
    m_request.httpStatus = extractStatusCode(iter->second);
    [...]
  }

  // Connection
  [...]

// we can check bool first, if it is true we don't need to .find()
  if (!hasCGIField && m_responseHeaders.find("content-type") != m_responseHeaders.end())
    hasCGIField = true;

  if (!hasCGIField && m_responseHeaders.find("location") != m_responseHeaders.end())
    hasCGIField = true;

  if (!hasCGIField) {
    LOG_ERROR << "No CGI field (Content-type | Status | Location) found";
    m_request.httpStatus = StatusInternalServerError;
    m_responseHeaders.clear(); // clear so we don't have additional headers in error response
    handleErrorBody();
  }
}
```

gwolf-011235 replied:
Check has case issue, you'd need toLower() headers first (which can also made const).

I suggest splitting the check into if no header whatsoever and if no cgi field present. Then we can give more info on what went wrong

u413-284-si replied:
The check receives the CGI body as is and expects the headers to have the correct case. Otherwise the header is not accepted/valid. Afterwards for processing the headers in the server the headers are lower cased in the following loop.

Split the check and created two errors specifying both cases.

gwolf-011235 replied:
Why though?

Header fields are case insensitive.

RFC 3875 Sect 6.3. [LINK](https://datatracker.ietf.org/doc/html/rfc3875#section-6.3): "The field-name is not case sensitive."

Same for "normal" headers in RFC 7230 Sect 3.2 [LINK](https://datatracker.ietf.org/doc/html/rfc7230#section-3.2): "Each header field consists of a case-insensitive field name [...]" - which is why we introduced toLower in the first place.

gwolf-011235 commented on src/ResponseBodyHandler.cpp:
```diff
@@ -108,6 +112,89 @@ void ResponseBodyHandler::execute()
 	}
 }
 
+void ResponseBodyHandler::parseCGIResponseBody()
```
parseCGIResponseBody() just encapsulates two functions (maybe one - see below) -> could be removed.

u413-284-si replied:
Removed

gwolf-011235 commented on src/ResponseBodyHandler.cpp:
```diff
@@ -108,6 +112,89 @@ void ResponseBodyHandler::execute()
+	const size_t sizeCRLFCRLF = 4;
+	size_t posHeadersEnd = m_responseBody.find("\r\n\r\n");
+
+	if (posHeadersEnd != std::string::npos) {
```
Here we could return early and print error message, if no header section is found:
```
if (posHeadersEnd == std::string::npos) {
  LOG_ERROR << "No header section found in CGI output";
  m_request.httpStatus = StatusInternalServerError;
  handleErrorBody();
}
```

u413-284-si replied:
Done

gwolf-011235 commented on src/ResponseBodyHandler.cpp:
```diff
@@ -108,6 +112,89 @@ void ResponseBodyHandler::execute()
+			lineStart = lineEnd + 2;
+			lineEnd = headers.find("\r\n", lineStart);
+		}
+		m_responseBody = m_responseBody.substr(posHeadersEnd + sizeCRLFCRLF);
```
I'm not sure if it makes a performance difference but .erase() would be semantically more correct for me

u413-284-si replied:
Changed, it does make a difference: substr requires reallocation of memory, erase works with existing memory. Doesnt make a difference in our case, but with large bodies it would

gwolf-011235 commented on src/ResponseBodyHandler.cpp:
```diff
@@ -108,6 +112,89 @@ void ResponseBodyHandler::execute()
+		m_request.httpStatus = extractStatusCode(iter->second);
+		if (m_request.httpStatus == StatusBadRequest) {
+			handleErrorBody();
+			m_responseHeaders.erase(iter);
```
I would m_responseHeaders.clear(): I think we don't want to print any header from the CGI output in case of error

u413-284-si replied:
We had a check in ResponseBuilder to only print from m_responseheaders if status < StatusMovedPermanently. But this way is better and doesnt necessitate the check and also the boolean _hasCGIStatusBadRequest_ introduced before

gwolf-011235 commented on src/ResponseBodyHandler.cpp:
```diff
@@ -108,6 +112,89 @@ void ResponseBodyHandler::execute()
+	if (iter != m_responseHeaders.end()) {
+		m_request.httpStatus = extractStatusCode(iter->second);
+		if (m_request.httpStatus == StatusBadRequest) {
+			handleErrorBody();
```
I think we should set status to StatusInternalServerError before entering here

u413-284-si replied:
Done

gwolf-011235 commented on src/ResponseBodyHandler.cpp:
```diff
@@ -108,6 +112,89 @@ void ResponseBodyHandler::execute()
+	iter = m_responseHeaders.find("connection");
+	if (iter != m_responseHeaders.end()) {
+		if (iter->second != "keep-alive" && iter->second != "close") {
+			m_request.httpStatus = StatusInternalServerError;
```
Like above: m_responseHeaders.clear()

u413-284-si replied:
Done

u413-284-si commented on inc/ResponseBodyHandler.hpp:
```diff
@@ -19,17 +19,22 @@
 	void handleErrorBody();
 	void setDefaultErrorPage();
+	void parseCGIResponseBody();
+	void parseCGIResponseStatusLine();
```
Removed

u413-284-si commented on inc/ResponseBuilder.hpp:
```diff
@@ -20,21 +23,22 @@
 public:
-	explicit ResponseBuilder(const FileSystemPolicy& fileSystemPolicy);
+	explicit ResponseBuilder(
+		const FileSystemPolicy& fileSystemPolicy, std::map<std::string, std::string>& responseHeaders);
```
Moved

gwolf-011235 reviewed (approved):
LGTM!

One P2: we still have `if (activeFd == connection.m_clientFd) return;` checks even though we shouldn't (?) need them

QCHR1581 commented on src/StatusCode.cpp:
```diff
@@ -0,0 +1,116 @@
+ * @param str The string representation of the HTTP status code.
+ * @return The corresponding statusCode enum value.
+ */
+statusCode stringToStatusCode(std::string& str)
```
I have also implemented a function to convert a string into a status code in #131. 

Gabriel saw, that my function has bug converting invalid values which are in between valid values. 
Therefore he recommended me to use your function, when this PR is merged. 

Unfortunately the return value ``StatusBadRequest`` does not fit my usecase for this function. When I use it to check the status code of the directive ``error_page``, 400 is a valid value. Which means, that I could not properly check, if the input string is valid or not. 

I would ask you, to change the return value of this function to ``NoStatus`` if the input string is not  a valid status code. Of course only, if this change does not have a too big impact.

u413-284-si replied:
Changed to NoStatus, have no objections against it. Makes handling actually easier

u413-284-si commented:
> LGTM!
> 
> One P2: we still have `if (activeFd == connection.m_clientFd) return;` checks even though we shouldn't (?) need them

In regards of the current connection that we are handling, it should not be required anymore as the clientFd is not active. But what about the case when we have several clients and there are other clientFds or cgiPipeEnds? I think we may need a different check à la: `if (activeFd != connection.m_pipeToCGIWriteEnd) return;`

gwolf-011235 commented:
I don't follow. 

Why should it make a difference if we have multiple clients? The problem was that the same connection would have upto 3 fds assigned. Since the status of the connection dictates which function we use a firing socket fd could call a cgi function (which is bad).

However a different client has its own connection object with its own status. Even if client1 is in status ReceiveCGI it doesn't bother client2 who is in status ReceiveBody

u413-284-si commented:
I had in mind that any active fd could trigger an event. Therefore I assumed that also fds from other connections could enter the CGI handling functions (connectionSendToCGI/connectionReceiveFromCGI). What I didn't have anymore in mind was that handleConnections() is connection-specific only passing on the fd of one particular connection. In that case we can omit not only the check for the CGI functions, but also connectionBuildResponse() and connectionSendResponse() as the CGI fds are closed after they are done.

gwolf-011235 commented:
Yes, I meant all checks.
